builtin/credential/aws/path_role.go

```diff
@@ -320,6 +322,64 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// updateUpgradableRoleEntries upgrades and persists all of the role entries
+// that are in need of being upgraded.
+func (b *backend) updateUpgradableRoleEntries(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// Upgrade only if we are either: (1) a local mount, or (2) are _not_ a
+	// performance replicated standby cluster.
+	if !(b.System().LocalMount() || !b.System().ReplicationState().HasState(
+		consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby)) {
+		return nil
+	}
+
+	// Read all the role names.
+	b.roleMutex.RLock()
+	roleNames, err := req.Storage.List(ctx, "role/")
+	b.roleMutex.RUnlock()
+	if err != nil {
+		return err
+	}
+
+	// Upgrade the roles as necessary.
+	for _, roleName := range roleNames {
+		err := b.updateUpgradableRoleEntry(ctx, req.Storage, roleName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updateUpgradableRoleEntry uses the write lock to read a role and persist it
+// if it needs to be upgraded.
+func (b *backend) updateUpgradableRoleEntry(ctx context.Context, s logical.Storage, roleName string) error {
+
+	b.roleMutex.Lock()
```

Comment:
Calling roleInternal will already upgrade the role if need be, so you don't need to manually upgrade. It also requires a write lock, not a read lock. We actually don't ever read lock so probably we should change it to a Mutex, not RWMutex.

Reply:
yep, I simplified that method so it just grabs the mutex and calls `roleInternal()`.  I didn't change the RWMutex to a Mutex yet, until we decide whether my aforementioned paranoia about running this in a goroutine warrants an RLock when reading the role names.

Reply:
changed to `Mutex'